src/runtime/pickle_runner.ts

```diff
@@ -395,9 +427,69 @@ export default class PickleRunner {
         },
       })
     }
-    return await this.invokeStep(
-      testStep.pickleStep,
-      testStep.stepDefinitions[0]
+    let stepResult
+    let afterStepHooksResult
+    const beforeStepHooksResult = await this.runStepHooks(
+      this.getBeforeStepHookDefinitions()
     )
+
+    if (beforeStepHooksResult.status !== Status.FAILED) {
+      stepResult = await this.invokeStep(
+        testStep.pickleStep,
+        testStep.stepDefinitions[0]
+      )
+      afterStepHooksResult = await this.runStepHooks(
+        this.getAfterStepHookDefinitions()
+      )
+    }
+    let cumulatedStepResult = beforeStepHooksResult
+
+    if (stepResult !== undefined) {
+      cumulatedStepResult = stepResult
+      if (beforeStepHooksResult?.duration !== null) {
+        cumulatedStepResult.duration = addDurations(
+          cumulatedStepResult.duration,
+          beforeStepHooksResult.duration
+        )
+      }
+      if (afterStepHooksResult?.duration !== null) {
+        cumulatedStepResult.duration = addDurations(
+          cumulatedStepResult.duration,
+          afterStepHooksResult.duration
+        )
+      }
+    }
+    return cumulatedStepResult
+  }
+
+  async runStepHooks(
+    stepHooks: TestStepHookDefinition[]
+  ): Promise<messages.TestStepFinished.ITestStepResult> {
+    const stepHooksResult = messages.TestStepFinished.TestStepResult.fromObject(
+      {
+        status:
+          this.result.status === Status.FAILED
+            ? Status.SKIPPED
+            : this.result.status,
+      }
```

Comment:
If we add a duration in here I think we can simplify code to avoid checking if duration is null both in this function and in the runStep function

Reply:
What exactly would I need to set it to? I tried:
        duration: this.result?.duration !== null ? this.result.duration : 0,
but this failed unit tests.

Reply:
In `src/time.ts` there is a method `getZeroDuration`. And then in places where you need to add durations you should no longer need null checks.